### ISSUE TITLE
Add methods zeroPosition and setPosition

### ISSUE THREAD

**Issue by damian99999** [pmarques-dev/PicoEncoder#4]
Hi,
I appreciate how this library works. The three encoders track position like a dream. The speed is much less noisy than other simple methods of deriving it. However, I am unable to easily zero the position. There is a workaround (another variable that keeps the position deviation at zero), but the real solution is to be able to set the internal encoder position to a known value. One of my encoders tracks a motorized axis that is rotating rapidly in one direction. I am afraid that after some time the internal position may overflow. I am not sure what the result will be. I hope this does not turn my machine into scrap. I prefer to reset the position when it is convenient. I think such a feature will be good for other users of this great library. Could you extend the class? I don't understand the PicoEncoder internals and i failed do do it myself.

**Comment by pmarques-dev** [comment]
Hi Damian,
It's great to ear some positive feedback from people actually using this library :)

My original plan was that people would just use something like this (note: this was written directly here on the comment, never seen a compiler):

static uint last_pos;
uint new_pos;
int delta;

new_pos = Encoder.position;
delta = new_pos - old_pos;
old_pos = new_pos;

This should always give you a "delta" from the previous position and is guaranteed to work correctly on overflow (two's complement arithmetic).

Having said this, I can understand this being tricky and not obvious to use, and I've had the feedback before from other users that they would prefer to read the position and reset it to zero, instead of computing differences, so I'll try to implement that, by popular demand ;)

The method needs to be something like "int get_and_reset_position(void);", though, as reading the position and resetting needs to be atomic and can not be two different operations, as you could lose steps in between reading and resetting.

By the way, if you're reading the motorized encoder position to compute the speed, you should be better of using the "speed" field in the PicoEncoder directly, as that is not exactly the same as computing position differences. The reason for that is not easy to explain, but it makes some difference at low speeds.

Basically you can think of it as: the "position" field has the best possible position estimation and the "speed" field has the best possible speed estimation, but since they don't use the exact same algorithm, you can't really get one from the other. The "speed" field should be easy to use with no overflow issues, though. It is also immune to sampling period jitter, which may be a plus depending on how you main loop looks like.

**Comment by damian99999** [comment]
You are  right about zeroing. C/C++ arithmetic on Pico run without exceptions on overflow so it will work flawlessly. If you can guarantee that there will be no glitch ("position" and "speed") on overflow then feature requested is irrelevant after all.

**Comment by pmarques-dev** [comment]
I wouldn't say irrelevant, as I'm bound to get that request again in the future if I don't implement it. I understand how a "get_and_reset" method can be easier to think about than worrying about overflows, etc., when you just want to periodically check how much the encoder has moved since last time.

For the position, the code above should give you the delta without any glitch on overflow (even when crossing the overflow back and forth, etc.). For the speed there is no overflow issue as the speed is just the most up to date speed estimate PicoEncoder can compute when update is called. So it can be positive or negative or zero, but it doesn't accumulate like the position does. You could only get overflow with an encoder going at more than 2³¹ / 64 -> ~33Msteps per second, but that's already over the 10Msteps per second max speed specification of PicoEncoder.

In any case, if you can use the delta method above for now, this becomes less urgent and I can wait to have a bit of free time to implement it. 

I also had another thing to implement on the backlog: incremental online calibration. Currently, if you want to calibrate phase sizes, you need to do that while the system is not being used, and save the calibration parameter to assign later. This creates an extra work that many won't be bothered to do, leaving the system with the default calibration which will give worse results, especially at low speeds. If we could call a calibration function as part of the system main loop and have PicoEncoder try to learn the phase sizes while operating, that could make the calibration simpler to use.

I'll try to implement these two features and release a new version at some point.
